pkg/reconciler/broker/create_brokercell.go

```diff
@@ -0,0 +1,89 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package broker
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/knative-gcp/pkg/reconciler/broker/resources"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/eventing/pkg/logging"
+	"knative.dev/eventing/pkg/reconciler/names"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/system"
+
+	brokerv1beta1 "github.com/google/knative-gcp/pkg/apis/broker/v1beta1"
+	inteventsv1alpha1 "github.com/google/knative-gcp/pkg/apis/intevents/v1alpha1"
+	brokercellresources "github.com/google/knative-gcp/pkg/reconciler/brokercell/resources"
+)
+
+// ensureBrokerCellExists creates a BrokerCell if it doesn't exist, and update broker status based on brokercell status.
+func (r *Reconciler) ensureBrokerCellExists(ctx context.Context, b *brokerv1beta1.Broker) error {
+	var bc *inteventsv1alpha1.BrokerCell
+	var err error
+	// TODO(#866) Get brokercell based on the label (or annotation) on the broker.
+	bc, err = r.brokerCellLister.BrokerCells(system.Namespace()).Get(resources.DefaultBroekrCellName)
+
+	if err != nil && !apierrs.IsNotFound(err) {
+		logging.FromContext(ctx).Error("Error reconciling brokercell", zap.String("namespace", b.Namespace), zap.String("broker", b.Name), zap.Error(err))
+		b.Status.MarkBrokerCelllUnknown("BrokerCellUnknown", "Failed to get brokercell %s/%s", bc.Namespace, bc.Name)
+		return err
+	}
+
+	if apierrs.IsNotFound(err) {
+		want := resources.CreateBrokerCell(b)
+		bc, err = r.RunClientSet.InternalV1alpha1().BrokerCells(want.Namespace).Create(want)
+		if err != nil && !apierrs.IsAlreadyExists(err) {
+			logging.FromContext(ctx).Error("Error creating brokercell", zap.String("namespace", b.Namespace), zap.String("broker", b.Name), zap.Error(err))
+			b.Status.MarkBrokerCelllFailed("BrokerCellCreationFailed", "Failed to create %s/%s", want.Namespace, want.Name)
+			return err
+		}
+		if apierrs.IsAlreadyExists(err) {
+			logging.FromContext(ctx).Info("Brokercell already exists", zap.String("namespace", b.Namespace), zap.String("broker", b.Name))
+			// There can be a race condition where the informer is not updated. In this case we directly
+			// read from the API server.
+			bc, err = r.RunClientSet.InternalV1alpha1().BrokerCells(want.Namespace).Get(want.Name, metav1.GetOptions{})
+			if err != nil {
+				logging.FromContext(ctx).Error("Failed to get the brokercell from the API server", zap.String("namespace", b.Namespace), zap.String("broker", b.Name), zap.Error(err))
+				return err
+			}
+		}
+		if err == nil {
+			r.Recorder.Eventf(b, corev1.EventTypeNormal, brokerCellCreated, "Created brokercell %s/%s", bc.Namespace, bc.Name)
```

Comment:
nit: created => ensured. Could be confusing if there is an existing brokercell.

Reply:
Actually I think `create` is accurate here: Line 53 tries to create a brokercell, and this line only records the event when there is no error.